### PR TITLE
[9.x] Flysystem v2

### PR DESCRIPTION
**Pull request by driesvints**
Initial attempt at migrating to (currently unreleased) Flysystem v2. Might be a little too soon for Laravel 8. In draft until Flysystem v2 has been released.

### Breaking Changes

- Any write operations (like `put`, `write`, `writeStream`, etc) now overwrites by default. Checking for existing files needs to be done in user land from now on.
- `writeStream` no longer throws a `FileExistsException`. Since this exception is nowhere used anymore in the framework, it's been removed.
- Similarly, `get` & `readStream` no longer throw a `FileNotFoundException` exception when a file exists. This is because the underlying Flysystem integration no longer distinguishes this specific failure from any other read failure. Both methods will return `null` when they cannot read a file.
- `delete` returns `true` now even if the file wasn't found.
- Using custom filesystem creators in a service provider now requires them to return an instance of `Illuminate\Contracts\Filesystem\Filesystem` rather than `League\Flysystem\FilesystemInterface` like before. See https://github.com/laravel/framework/pull/33612#discussion_r460018589
- The FTP driver now ships as a separate league package and needs to be explicitly required through Composer for it to be used.
- The concept of the cached adapter has been removed from Flysystem v2 and thus also from Laravel.
- `Illuminate\Filesystem\FilesystemAdapter`'s constructor now requires an extra `$adapter` argument and an optional `$options` argument.

### CODE REVIEW

File: src/Illuminate/Filesystem/FilesystemManager.php [hunk 1]
```diff
@@ -140,13 +143,7 @@ protected function resolve($name)
...
     protected function callCustomCreator(array $config)
     {
-        $driver = $this->customCreators[$config['driver']]($this->app, $config);
-
-        if ($driver instanceof FilesystemInterface) {
-            return $this->adapt($driver);
-        }
-
-        return $driver;
+        return $this->customCreators[$config['driver']]($this->app, $config);
```
Comment (driesvints) [inline src/Illuminate/Filesystem/FilesystemManager.php:146]: These changes require custom creators to return an instance of `Illuminate\Contracts\Filesystem\Filesystem` rather than an instance of `League\Flysystem\FilesystemInterface`. The reason for that is that creating a `FilesystemAdapter` instance is now much more complex then what the previous `adapt` method did (see a bit further below).

**Comment by driesvints** [comment]
Feel free to ignore the failing "prefer-lowest" tests as they test the old alpha.1 which get pulled in because of composer's "prefer-lowest" flag. The most recent version works.

File: composer.json [hunk 2]
```diff
@@ -82,7 +82,9 @@
         "doctrine/dbal": "^2.6",
         "filp/whoops": "^2.4",
         "guzzlehttp/guzzle": "^6.5.5|^7.0.1",
-        "league/flysystem-cached-adapter": "^1.0",
+        "league/flysystem-aws-s3-v3": "^2.0",
+        "league/flysystem-ftp": "^2.0.0-alpha.4|^2.0",
```
Comment (Seldaek) [inline composer.json:86]: `^2.0.0-alpha4` is enough, the `|^2.0` is included in that range and redundant.

Comment (Seldaek) [reply composer.json:86]: Well, technically ^2.0 includes even more versions, as I just tweeted, so actually it needs to be gone otherwise it'll still allow alpha1

Comment (driesvints) [reply composer.json:86]: Right. Good catch 👍 

I think ultimately I was trying to find a way to keep the "^2.0" constraint so the PR would be "ready to go" as soon as Flysystem v2 got released but I guess that's not possible as long as it's unreleased. Anyway, thanks for helping out!

Comment (Seldaek) [reply composer.json]: It is ready to go, `^2.0-alpha4` is completely valid actually more accurate and correct than ^2.0 (which again resolves to `^2.0-dev` internally). It just looks odd because we don't see a lot of these, but it's not wrong.

Comment (driesvints) [reply composer.json]: That's true. Hadn't thought of it like that yet.

File: composer.json [hunk 3]
```diff
@@ -82,7 +82,9 @@
         "doctrine/dbal": "^2.6|^3.0",
         "filp/whoops": "^2.8",
         "guzzlehttp/guzzle": "^7.2",
-        "league/flysystem-cached-adapter": "^1.0",
+        "league/flysystem-aws-s3-v3": "^2.0",
+        "league/flysystem-ftp": "^2.0.0-RC1",
```
Comment (GrahamCampbell) [inline composer.json:86]: Why not `^2.0`?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: composer.json (hunk 3 above)
Comment (driesvints) [reply composer.json:86]: See https://github.com/laravel/framework/pull/33612#discussion_r465021969

Comment (GrahamCampbell) [reply composer.json:86]: That comment no longer applies now that 2.0.0 is tagged.

Comment (driesvints) [reply composer.json]: ah correct!

File: src/Illuminate/Contracts/Filesystem/Filesystem.php [hunk 4]
```diff
@@ -30,9 +30,7 @@ public function exists($path);
      * Get the contents of a file.
      *
      * @param  string  $path
-     * @return string
-     *
-     * @throws \Illuminate\Contracts\Filesystem\FileNotFoundException
+     * @return string|null
```
Comment (GrahamCampbell) [inline src/Illuminate/Contracts/Filesystem/Filesystem.php:33]: Why the change in interface? Could we not continue to throw the exception, so upgrading is easier for everyone?

File: src/Illuminate/Contracts/Filesystem/Filesystem.php [hunk 5]
```diff
@@ -41,8 +39,6 @@ public function get($path);
      *
      * @param  string  $path
      * @return resource|null The path resource or null on failure.
-     *
-     * @throws \Illuminate\Contracts\Filesystem\FileNotFoundException
```
Comment (GrahamCampbell) [inline src/Illuminate/Contracts/Filesystem/Filesystem.php:45]: Similar here. Perhaps throw an exception instead of null for the other kind of failure too?

File: src/Illuminate/Contracts/Filesystem/Filesystem.php [hunk 6]
```diff
@@ -63,9 +59,6 @@ public function put($path, $contents, $options = []);
      * @param  resource  $resource
      * @param  array  $options
      * @return bool
-     *
-     * @throws \InvalidArgumentException If $resource is not a file handle.
-     * @throws \Illuminate\Contracts\Filesystem\FileExistsException
```
Comment (GrahamCampbell) [inline src/Illuminate/Contracts/Filesystem/Filesystem.php:68]: What happens on failure now? Just returns false? What about the invalid arg exception from before? Seems odd to no longer be able to tell the difference between a failed write and the code being invoked wrong by a dev?

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Contracts/Filesystem/Filesystem.php (hunk 4 above)
Comment (driesvints) [reply src/Illuminate/Contracts/Filesystem/Filesystem.php:33]: I've explained this in the main comment of the PR.

File: src/Illuminate/Contracts/Filesystem/Filesystem.php (hunk 5 above)
Comment (driesvints) [reply src/Illuminate/Contracts/Filesystem/Filesystem.php:45]: I've explained this in the main comment of the PR.

File: src/Illuminate/Contracts/Filesystem/Filesystem.php (hunk 6 above)
Comment (driesvints) [reply src/Illuminate/Contracts/Filesystem/Filesystem.php:68]: I've explained this in the main comment of the PR.

File: src/Illuminate/Contracts/Filesystem/Filesystem.php (hunk 4 above)
Comment (GrahamCampbell) [reply src/Illuminate/Contracts/Filesystem/Filesystem.php:33]: You didn't explain why though, just that it happened? Why do we need this breaking change?

**Review by GrahamCampbell — APPROVED** [review]

**Comment by taylorotwell** [comment]
I need quite a bit more info on:

> Illuminate\Filesystem\FilesystemAdapter's constructor now requires an extra $adapter argument and an optional $options argument.

I have to document more info than this - what should I include specifically? Can you give me a before and after example of something that might affect user land code?

> Using custom filesystem creators in a service provider now requires them to return an instance of Illuminate\Contracts\Filesystem\Filesystem rather than League\Flysystem\FilesystemInterface like before. See [9.x] Flysystem v2 #33612 (comment)

Does this mean all custom disk implementations are now broken? If so, can you provide a before and after example of how to fix a hypothetical custom disk so that I can properly document the process in the upgrade guide. Otherwise, users will have nothing to go on if all I mention is the sentence above.

**Comment by driesvints** [comment]
> I have to document more info than this - what should I include specifically? Can you give me a before and after example of something that might affect user land code?

No real userland code is affected here, only when people are instantiating their own filesystem adapters or packages adding new drivers.

```php
use Illuminate\Filesystem\FilesystemAdapter;

$flysystem = new Flysystem($adapter, $config);

// Before...
$filesystem = new FilesystemAdapter($flysystem);

// After...
return new FilesystemAdapter($flysystem, $adapter, $config);
```

> Does this mean all custom disk implementations are now broken? If so, can you provide a before and after example of how to fix a hypothetical custom disk so that I can properly document the process in the upgrade guide. Otherwise, users will have nothing to go on if all I mention is the sentence above. In other words, can you give me an example of how to define a custom disk in Laravel 8.x (there is one in docs) and then an example of how to do it in Laravel 9.x?

Yes, custom disks will need to update their implementation.

```php
// Before...
use Illuminate\Support\Facades\Storage;
use League\Flysystem\Filesystem as Flysystem;
use Spatie\Dropbox\Client as DropboxClient;
use Spatie\FlysystemDropbox\DropboxAdapter;

Storage::extend('dropbox', function ($app, $config) {
    $client = new DropboxClient(
        $config['authorization_token']
    );

    return new Filesystem(new DropboxAdapter($client));
});

// After...
use Illuminate\Filesystem\FilesystemAdapter;
use Illuminate\Support\Facades\Storage;
use League\Flysystem\Filesystem;
use Spatie\Dropbox\Client as DropboxClient;
use Spatie\FlysystemDropbox\DropboxAdapter;

Storage::extend('dropbox', function ($app, $config) {
    $client = new DropboxClient(
        $config['authorization_token']
    );

    $adapter = new DropboxAdapter($client);
    $flysystem = new Flysystem($adapter, $config);

    return new FilesystemAdapter($flysystem, $adapter, $config);
});
```
